Zend/zend_language_parser.y

```diff
@@ -798,6 +798,7 @@ type_expr:
 	|	'?' type			{ $$ = $2; $$->attr |= ZEND_TYPE_NULLABLE; }
 	|	union_type			{ $$ = $1; }
 	|	intersection_type	{ $$ = $1; }
+	|	'?' intersection_type	{ $$ = $2; $$->attr |= ZEND_TYPE_NULLABLE; }
```

Comment:
Regarding the nullability syntax.  I find `?X&Y`, while technically unambiguous due to `null&Y` not making sense, just isn't great for readability.

Further, we decided against `?X|Y` in the union types RFC, and I think that decision should hold (even though the reasons are somewhat different.

I've got an implementation for `(X&Y)|null`, which is my preferred syntax at: https://gist.github.com/sgolemon/660b105e6894a6341792a479cda8f76b

However, we are much too late for 8.1 at this point and syntax is too important to rush, so this is going to have to wait for 8.2 no matter what we do.

Reply:
> Regarding the nullability syntax. I find ?X&Y, while technically unambiguous due to null&Y not making sense, just isn't great for readability.

> Further, we decided against ?X|Y in the union types RFC, and I think that decision should hold (even though the reasons are somewhat different.

Agreed.

> I've got an implementation for `(X&Y)|null`

I'm not opposed to allowing parenthesis, but requiring them seems unnecessary - the precedence of `&`/`|` and `&&`/`||` is known in other contexts in php (operator precedence).

If `(A&B)|null` was allowed, there'd be the question of whether `(A)|null` should be allowed, `(A)|(null)`, etc. It seems like it should (IF allowed), though the syntax seems less readable

> However, we are much too late for 8.1 at this point and syntax is too important to rush, so this is going to have to wait for 8.2 no matter what we do.

Agreed.

Reply:
I don't think you can use `?X|Y` being disallowed to conclude that everyone would also be against the special case of `?X&Y`. At the time in the Github discussion I argued against `?(X|Y)`, because it added parenthesis to not be ambiguous which would otherwise not be needed, and `X|Y|null` was available as an alternative. In that case having a more complicated syntax only to somehow include `?` would have been counterproductive, in my opinion.

Yet I very much see the value of `?X&Y` - they enable optional parameters/properties, and they do not contain parenthesis. If combined union type and intersection type syntax ever makes it into PHP, it would be easy to not allow `?` there, just as `?` is not allowed for union types. So the syntax seems forward-compatible and geared towards a common use case in PHP.

Reply:
`?` + not nullable simple/complex type expr. is clearly unambiguous grammar (`A&?B` or `A|(?B)` and other complex usecases are not changed by this PR, it can be seen like a syntactical sugar even if parenthesis later would be required for mixed union/insersection types)

Reply:
> If (A&B)|null was allowed, there'd be the question of whether (A)|null should be allowed, (A)|(null), etc. It seems like it should (IF allowed), though the syntax seems less readable

That's a fair complaint.  For me it comes down to finding a balance between providing the functionality that's being asked for quite strongly, and not over-delivering syntax that hasn't been entirely thought through.  That's why my version of this patch only allows `null` on the right hand side, and requires parens around the intersection. It's the MOST restrictive syntax, and thereby the lease likely to cause issues after we've spent more time thinking about it.